guppylang/checker/expr_checker.py

```diff
@@ -1138,25 +1151,43 @@ def eval_py_expr(node: PyExpr, ctx: Context) -> Any:
     return python_val
 
 
-def python_value_to_guppy_type(v: Any, node: ast.expr, globals: Globals) -> Type | None:
+def python_value_to_guppy_type(
+    v: Any, node: ast.expr, globals: Globals, type_hint: Type | None = None
+) -> Type | None:
     """Turns a primitive Python value into a Guppy type.
 
+    Accepts an optional `type_hint` for the expected expression type that is used to
```

Comment:
Currently it's not an internal error, the hint is just ignored if it's invalid. The actual checking that the synthesised type matches what's expected happens in the `ExprChecker`.

A wider refactor of this would be good, but imo too much for this PR.